jlm/rvsdg/traverser.cpp

```diff
@@ -5,293 +5,235 @@
  */
 
 #include <jlm/rvsdg/graph.hpp>
-#include <jlm/rvsdg/notifiers.hpp>
 #include <jlm/rvsdg/traverser.hpp>
 
-using namespace std::placeholders;
-
-/* top down traverser */
-
 namespace jlm::rvsdg
 {
 
-TopDownTraverser::~TopDownTraverser() noexcept = default;
-
-TopDownTraverser::TopDownTraverser(Region * region)
-    : region_(*region)
+namespace detail
 {
-  for (auto & node : region->TopNodes())
-    tracker_.set_nodestate(&node, traversal_nodestate::frontier);
+template<bool IsConst>
+TopDownTraverserGeneric<IsConst>::~TopDownTraverserGeneric() noexcept = default;
 
-  for (size_t n = 0; n < region->narguments(); n++)
+template<bool IsConst>
+TopDownTraverserGeneric<IsConst>::TopDownTraverserGeneric(RegionType * region)
+{
+  for (auto & node : region->Nodes())
   {
-    auto argument = region->argument(n);
-    for (const auto & user : argument->Users())
+    unvisited_.Insert(&node);
+
+    // Build the initial frontier from nodes that have no predecessor nodes
+    bool onFrontier = true;
+    for (auto & input : node.Inputs())
     {
-      if (auto node = TryGetOwnerNode<Node>(user))
+      if (TryGetOwnerNode<Node>(*input.origin()))
       {
-        if (!predecessors_visited(node))
-          continue;
-
-        tracker_.set_nodestate(node, traversal_nodestate::frontier);
+        onFrontier = false;
+        break;
       }
     }
+    if (onFrontier)
+      frontier_.emplace(node.GetNodeId(), &node);
   }
-
-  callbacks_.push_back(on_node_create.connect(std::bind(&TopDownTraverser::node_create, this, _1)));
-  callbacks_.push_back(
-      on_input_change.connect(std::bind(&TopDownTraverser::input_change, this, _1, _2, _3)));
 }
 
+template<bool IsConst>
 bool
-TopDownTraverser::predecessors_visited(const Node * node) noexcept
+TopDownTraverserGeneric<IsConst>::allPredecessorsVisited(const Node & node) const noexcept
 {
-  for (size_t n = 0; n < node->ninputs(); n++)
+  for (auto & input : node.Inputs())
   {
-    auto predecessor = TryGetOwnerNode<Node>(*node->input(n)->origin());
-    if (!predecessor)
-      continue;
-
-    if (tracker_.get_nodestate(predecessor) != traversal_nodestate::behind)
+    if (auto predecessor = TryGetOwnerNode<Node>(*input.origin());
+        predecessor && unvisited_.Contains(predecessor))
       return false;
   }
 
   return true;
 }
 
-Node *
-TopDownTraverser::next()
+template<bool IsConst>
+bool
+TopDownTraverserGeneric<IsConst>::prepareFrontier()
 {
-  Node * node = tracker_.peek();
-  if (!node)
-    return nullptr;
+  if (!frontier_.empty())
+    return true;
 
-  tracker_.set_nodestate(node, traversal_nodestate::behind);
-  for (size_t n = 0; n < node->noutputs(); n++)
+  if (unvisited_.IsEmpty())
+    return false;
+
+  // We still have unvisited nodes, despite the frontier being empty
+  // This can happen in rare cases when nodes ahead of the frontier change their inputs.
+  // Go over the unvisited nodes again to create a new frontier
+  for (auto & node : unvisited_.Items())
```

Comment:
Well no, this means you can not just keep track of the visited nodes, and need the unvisited nodes.